nomos-services/key-management-system-macros/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+    let public_key_encoding_enum =
+        build_encoding_enum(public_key_encoding_type_ident, key_enum_variants);
+
+    quote! {
+        #payload_encoding_enum
+        #signature_encoding_enum
+        #public_key_encoding_enum
+    }
+}
+
+fn build_encoding_enum(
+    encoding_type: &Ident,
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let encoding_enum_ident = Ident::new(&format!("{encoding_type}Encoding"), Span::call_site());
+    let encoding_enum_variants = key_enum_variants
+        .iter()
+        .map(|variant| build_encoding_enum_variant(encoding_type, variant));
+
+    quote! {
+        pub enum #encoding_enum_ident {
+            #(#encoding_enum_variants),*
+        }
+    }
+}
+
+fn build_encoding_enum_variant(encoding_type: &Ident, key_enum_variant: &Variant) -> Variant {
+    let variant_ident = key_enum_variant.ident.clone();
+    let variant_attributes_cfg: Vec<Attribute> =
+        get_cfg_attributes(&key_enum_variant.attrs).collect();
+
+    let wrapped_key = get_wrapped_key_as_ref(key_enum_variant);
+    let wrapped_key_type = &wrapped_key.ty;
+    let wrapped_key_encoding_type: Type =
+        parse_quote!(<#wrapped_key_type as crate::keys::secured_key::SecuredKey>::#encoding_type);
+
+    let wrapped_type_field = Field {
+        attrs: vec![],
+        vis: Visibility::Inherited,
+        mutability: FieldMutability::None,
+        ident: None,
+        colon_token: None,
+        ty: wrapped_key_encoding_type,
+    };
+    let unnamed = Punctuated::<Field, _>::from_iter(vec![Pair::End(wrapped_type_field)]);
+    let field = Fields::Unnamed(FieldsUnnamed {
+        paren_token: Paren::default(),
+        unnamed,
+    });
+    Variant {
+        attrs: variant_attributes_cfg,
+        ident: variant_ident,
+        fields: field,
+        discriminant: None,
+    }
+}
+
+fn build_key_enum_impl_secured_key(
+    ident: &Ident,
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let key_enum_impl_secured_key_method_sign =
+        build_key_enum_impl_secured_key_method_sign(key_enum_variants);
+    let key_enum_impl_secured_key_method_as_public_key =
+        build_key_enum_impl_secured_key_method_as_public_key(key_enum_variants);
+
+    quote! {
+        impl crate::keys::secured_key::SecuredKey for #ident
+        {
+            type Payload = PayloadEncoding;
+            type Signature = SignatureEncoding;
+            type PublicKey = PublicKeyEncoding;
+            type Error = crate::keys::errors::KeyError;
+
+            #key_enum_impl_secured_key_method_sign
+            #key_enum_impl_secured_key_method_as_public_key
+        }
+    }
+}
+
+fn build_key_enum_impl_secured_key_method_sign(
+    key_enum_variants: &Punctuated<Variant, Comma>,
+) -> TokenStream {
+    let match_block = if key_enum_variants.is_empty() {
+        quote! {
+            Err(Self::Error::NoKeysEnabled)
+        }
```

Comment:
Is this so it can be used in an empty enum?

Reply:
Correct. 
I'll give the full answer in the thread above.